enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala

```diff
@@ -0,0 +1,92 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime.slotted.pipes
+
+import org.neo4j.cypher.internal.compatibility.v3_4.runtime.{Slot, SlotConfiguration}
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+import org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.Expression
+import org.neo4j.cypher.internal.runtime.interpreted.pipes.{Pipe, PipeWithSource, QueryState}
+import org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext
+import org.neo4j.cypher.internal.runtime.slotted.helpers.SlottedPipeBuilderUtils
+import org.neo4j.cypher.internal.runtime.{LongArraySet, PrefetchingIterator}
+import org.neo4j.cypher.internal.util.v3_4.attribution.Id
+
+import scala.collection.immutable
+
+case class DistinctSlottedPrimitivePipe(source: Pipe,
+                                        slots: SlotConfiguration,
+                                        primitiveSlots: Array[Int],
+                                        projections: Map[Slot, Expression])
+                                       (val id: Id = Id.INVALID_ID)
+  extends PipeWithSource(source) {
+
+  //===========================================================================
+  // Compile-time initializations
+  //===========================================================================
+  private val setValuesInOutput: immutable.Iterable[(ExecutionContext, QueryState, ExecutionContext) => Unit] = projections.map {
+    case (slot, expression) =>
+      val f = SlottedPipeBuilderUtils.makeSetValueInSlotFunctionFor(slot)
+      (incomingContext: ExecutionContext, state: QueryState, outgoingContext: ExecutionContext) =>
+        f(outgoingContext, expression(incomingContext, state))
+  }
+
+  projections.values.foreach(_.registerOwningPipe(this))
+
+  //===========================================================================
+  // Runtime code
+  //===========================================================================
+  protected def internalCreateResults(input: Iterator[ExecutionContext],
+                                      state: QueryState): Iterator[ExecutionContext] = {
+    new PrefetchingIterator[ExecutionContext] {
+      private val seen = new LongArraySet(32, projections.size)
+
+      override def produceNext(): Option[ExecutionContext] = {
+        while (input.nonEmpty) {
+          val next: ExecutionContext = input.next()
+
+          // Create key array
+          val keys = buildKey(next)
+
+          if (seen.add(keys)) {
+            // Found something! Set it as the next element to yield, and exit
+            val outgoing = SlottedExecutionContext(slots)
+            for (setter <- setValuesInOutput) {
+              setter(next, state, outgoing)
+            }
+
+            return Some(outgoing)
+          }
+        }
+
+        None
+      }
```

Comment:
If you have measurement indicating that this way of _iterating_ is faster than the old way in `DistinctSlottedPipe` it would be good to apply it to that too.

Reply:
Hmm... Thanks for pointing this out. Not sure I have any reasons to prefer this... Looks like the 
builtin Scala collection `filter` does something very similar... I'll double check

Reply:
I understand why I did this now. In the old `DistinctSlottedPipe`, we build the output execution context and execute the expressions on that context. Here, the "distinctification" is done before creating the output `ExecutionContext`. I don't think this pattern would be applicable for the old pipe.